File: cmux-tui/bindings/typescript/e2e/e2e.ts
import {
  CmuxClient,
  CmuxCommandError,
  CmuxTimeoutError,
  Tree,
  UUID,
} from "../src/index.js";

async function main(): Promise<void> {
  const socketPath = process.env.CMUX_TUI_SOCKET || process.env.CMUX_MUX_SOCKET;
  if (!socketPath) throw new Error("CMUX_TUI_SOCKET is required");

  const marker = `CMUX_TS_E2E_${process.pid}_${Date.now()}`;
  const later = `${marker}_ATTACH`;
  const client = new CmuxClient({ socketPath, timeoutMs: 5000 });
  try {
    const identify = await client.identify();
    assert(identify.app === "cmux-tui", `unexpected app ${identify.app}`);
    assert(identify.protocol >= 5 && identify.protocol <= 8, `unsupported protocol ${identify.protocol}`);
    assert(
      identify.protocol >= 8 && identify.capabilities?.includes("topology-resume-v1"),
      "server omitted protocol-v8 topology capabilities",
    );
    assert(identify.canonical_topology_revision !== undefined, "identify omitted canonical revision");
    const ping = await client.ping();
    assert(ping.ok, "ping reported false");
    assert(ping.session_id === identify.session_id, "ping session authority changed");
    assert(ping.daemon_instance_id === identify.daemon_instance_id, "ping daemon authority changed");
    assert(ping.topology_revision === identify.topology_revision, "ping legacy revision changed");
    assert(
      ping.canonical_topology_revision === identify.canonical_topology_revision,
      "ping canonical revision changed",
    );

    const created = await client.newWorkspace({ name: marker, cols: 80, rows: 24 });
    await client.send(created.surface, { text: `printf '${marker}\\n'\r` });
    await waitForMarker(client, created.surface, marker);
    const screen = await client.readScreen(created.surface);
    assert(screen.text.includes(marker), "marker missing from read-screen");

    const tree = await client.listWorkspaces();
    const workspaceId = findWorkspaceForSurface(tree, created.surface);
    assert(workspaceId !== undefined, "new workspace not found");

    const snapshot = await client.topologySnapshot();
    const canonical = snapshot.topology.workspaces.find((workspace) =>
      workspace.screens.some((screen) =>
        screen.panes.some((pane) => pane.tabs.some((tab) => tab.id === created.surface))));
    assert(canonical?.id === workspaceId, "canonical topology omitted the new workspace");
    const topology = await client.subscribeTopology({
      daemon_instance_id: snapshot.daemon_instance_id,
      session_id: snapshot.session_id,
      revision: snapshot.revision,
    });
    assert(topology.status === "subscribed", `fresh snapshot required resnapshot: ${JSON.stringify(topology)}`);
    await client.renameWorkspace(workspaceId!, `${marker}-topology`);
    const topologyEvent = await stage("topology rename delta", topology.stream.next(2000));
    assert(topologyEvent.event === "topology-delta", `unexpected topology event ${JSON.stringify(topologyEvent)}`);
    if (topologyEvent.event === "topology-delta") {
      assert(topologyEvent.operation === "workspace-renamed", "wrong topology operation");
      assert(topologyEvent.base_revision === snapshot.revision, "wrong topology base revision");
      assert(topologyEvent.revision === snapshot.revision + 1, "wrong topology revision");
    }
    topology.stream.close();
    const stale = await client.subscribeTopology({
      daemon_instance_id: "00000000-0000-0000-0000-000000000001" as UUID,
      session_id: snapshot.session_id,
      revision: snapshot.revision,
    });
    assert(stale.status === "resnapshot-required", "stale daemon cursor unexpectedly subscribed");
    if (stale.status === "resnapshot-required") {
      assert(stale.reason === "stale-daemon", `wrong stale cursor reason ${stale.reason}`);
    }

    const events = await client.subscribe();
    const title = `${marker}-title`;
    await client.send(created.surface, { text: `printf '\\033]2;${title}\\007'; sleep 30\r` });
    const titleChanged = await stage(
      "OSC title event",
      nextTitleChanged(events, created.surface, title, 3000),
    );
    assert(titleChanged.title === title, `bad title event ${JSON.stringify(titleChanged)}`);
    await client.send(created.surface, { text: "\x03" });
    await client.resizeSurface(created.surface, 100, 31);
    const resized = await stage(
      "surface resize event",
      nextSurfaceResized(events, created.surface, 1000),
    );
    assert(resized.cols === 100 && resized.rows === 31, `bad resize event ${JSON.stringify(resized)}`);
    await client.resizeSurface(created.surface, 100, 31);
    const duplicate = await nextSurfaceResized(events, created.surface, 500).catch((err) => {
      if (err instanceof CmuxTimeoutError) return null;
      throw err;
    });
    assert(duplicate === null, "same-size resize emitted surface-resized");
    events.close();
    await client.renameSurface(created.surface, `${marker}-renamed`);

    const attach = await client.attachSurface(created.surface);
    const first = await stage("initial attach state", attach.next(1000));
    assert(first.event === "vt-state", `first attach event was ${first.event}`);
    await client.send(created.surface, { text: `printf '${later}\\n'\r` });
    const output = await stage("attach output", nextAttachOutput(attach, 3000));
    assert(output.event === "output" || output.event === "resized", "attach did not produce output/resized after vt-state");
    attach.close();

    await client.closeWorkspace(workspaceId!);
    const afterClose = await client.listWorkspaces();
    assert(findWorkspaceForSurface(afterClose, created.surface) === undefined, "closed workspace still present");
    try {
      await client.readScreen(created.surface);
      throw new Error("read-screen on closed surface unexpectedly succeeded");
    } catch (err) {
      assert(err instanceof CmuxCommandError, `closed surface error was not command error: ${err}`);
      assert(String(err.message).length > 0, "command error did not preserve server message");
    }
  } finally {
    await client.close().catch(() => undefined);
  }
}

async function nextTitleChanged(
  events: Awaited<ReturnType<CmuxClient["subscribe"]>>,
  surface: number,
  title: string,
  timeoutMs: number,
): Promise<{ event: "title-changed"; surface: number; title: string }> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const event = await events.next(Math.max(1, deadline - Date.now()));
    if (
      event.event === "title-changed" &&
      "surface" in event &&
      event.surface === surface &&
      "title" in event &&
      event.title === title
    ) {
      return { event: "title-changed", surface: event.surface, title: event.title };
    }
  }
  throw new Error("title-changed event not observed");
}

async function waitForMarker(client: CmuxClient, surface: number, marker: string): Promise<void> {
  const deadline = Date.now() + 5000;
  let last = "";
  while (Date.now() < deadline) {
    last = (await client.readScreen(surface)).text;
    if (last.includes(marker)) return;
    await new Promise((resolve) => setTimeout(resolve, 50));
  }
  throw new Error(`marker not found; last screen: ${JSON.stringify(last)}`);
}

async function nextSurfaceResized(events: Awaited<ReturnType<CmuxClient["subscribe"]>>, surface: number, timeoutMs: number) {
  const deadline = Date.now() + timeoutMs;
  for (;;) {
    const remaining = deadline - Date.now();
    if (remaining <= 0) throw new CmuxTimeoutError("surface-resized not observed");
    const event = await events.next(remaining);
    if (event.event === "surface-resized" && event.surface === surface) return event;
  }
}

async function nextAttachOutput(attach: Awaited<ReturnType<CmuxClient["attachSurface"]>>, timeoutMs: number) {
  const deadline = Date.now() + timeoutMs;
  for (;;) {
    const remaining = deadline - Date.now();
    if (remaining <= 0) throw new CmuxTimeoutError("attach output not observed");
    const event = await attach.next(remaining);
    if (event.event === "output" || event.event === "resized") return event;
  }
}

function findWorkspaceForSurface(tree: Tree, surface: number): number | undefined {
  for (const workspace of tree.workspaces) {
    for (const screen of workspace.screens) {
      for (const pane of screen.panes) {
        if ("tabs" in pane && pane.tabs?.some((tab) => tab.surface === surface)) return workspace.id;
      }
    }
  }
  return undefined;
}

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) throw new Error(message);
}

async function stage<T>(name: string, operation: Promise<T>): Promise<T> {
  try {
    return await operation;
  } catch (error) {
    throw new Error(`${name}: ${error instanceof Error ? error.message : String(error)}`);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: cmux-tui/bindings/python/e2e.py
#!/usr/bin/env python3
from __future__ import annotations

import os
import sys
import time
from pathlib import Path
from uuid import UUID

sys.path.insert(0, str(Path(__file__).resolve().parent))

from cmux import (  # noqa: E402
    CommandError,
    CmuxClient,
    TimeoutError as CmuxTimeoutError,
    TopologyCursor,
    TopologyDelta,
    TopologyOperation,
    TopologyResnapshotReason,
    TopologyResnapshotRequired,
    TopologyStream,
)


def main() -> int:
    socket_path = os.environ.get("CMUX_TUI_SOCKET") or os.environ.get("CMUX_MUX_SOCKET")
    if not socket_path:
        raise SystemExit("CMUX_TUI_SOCKET is required")
    marker = f"CMUX_PY_E2E_{os.getpid()}_{time.time_ns()}"
    later = f"{marker}_ATTACH"
    with CmuxClient(socket_path=socket_path, timeout=5.0, allow_protocol_v6_attach=True) as client:
        info = client.identify()
        assert info.app == "cmux-tui", info
        assert 5 <= info.protocol <= 8, info
        assert info.supports_topology_v8, info
        assert info.topology_cursor is not None, info
        ping = client.ping()
        assert ping.ok, ping
        assert ping.session_id == info.session_id, (ping, info)
        assert ping.daemon_instance_id == info.daemon_instance_id, (ping, info)
        assert ping.topology_revision == info.topology_revision, (ping, info)
        assert ping.canonical_topology_revision == info.topology_cursor.revision, (ping, info)
        created = client.new_workspace(name=marker, cols=80, rows=24)
        client.send(created.surface, text=f"printf '{marker}\\n'\r")
        wait_for_marker(client, created.surface, marker)
        assert marker in client.read_screen(created.surface).text
        workspace = find_workspace_for_surface(client.list_workspaces(), created.surface)
        assert workspace is not None
        snapshot = client.topology_snapshot()
        canonical = next(
            item
            for item in snapshot.topology.workspaces
            if any(
                tab.id == created.surface
                for screen in item.screens
                for pane in screen.panes
                for tab in pane.tabs
            )
        )
        assert canonical.id == workspace
        topology = client.subscribe_topology(snapshot.cursor)
        assert isinstance(topology, TopologyStream), topology
        client.rename_workspace(workspace, f"{marker}-topology")
        delta = next_topology(topology, 2.0)
        assert isinstance(delta, TopologyDelta), delta
        assert delta.operation == TopologyOperation.WORKSPACE_RENAMED
        assert delta.base_revision == snapshot.revision
        assert delta.revision == snapshot.revision + 1
        topology.close()
        stale = client.subscribe_topology(
            TopologyCursor(
                UUID("00000000-0000-0000-0000-000000000001"),
                snapshot.session_id,
                snapshot.revision,
            )
        )
        assert isinstance(stale, TopologyResnapshotRequired), stale
        assert stale.reason == TopologyResnapshotReason.STALE_DAEMON
        events = client.subscribe()
        try:
            title = f"{marker}_TITLE"
            client.send(created.surface, text=f"printf '\\033]2;{title}\\007'; sleep 30\r")
            title_changed = next_title_changed(events, created.surface, title, 3.0)
            assert title_changed.title == title, title_changed
            client.send(created.surface, text="\x03")
            client.resize_surface(created.surface, 100, 31)
            resized = next_resized(events, created.surface, 1.0)
            assert (resized.cols, resized.rows) == (100, 31)
            client.resize_surface(created.surface, 100, 31)
            try:
                next_resized(events, created.surface, 0.5)
            except CmuxTimeoutError:
                pass
            else:
                raise AssertionError("same-size resize emitted surface-resized")
        finally:
            events.close()
        client.rename_surface(created.surface, f"{marker}-renamed")
        attach = client.attach_surface(created.surface)
        try:
            first = next(attach)
            assert first.event == "vt-state", first
            client.send(created.surface, text=f"printf '{later}\\n'\r")
            next_attach_output(attach, 3.0)
        finally:
            attach.close()
        client.close_workspace(workspace)
        assert find_workspace_for_surface(client.list_workspaces(), created.surface) is None
        try:
            client.read_screen(created.surface)
        except CommandError as exc:
            assert exc.message
        else:
            raise AssertionError("read-screen on closed surface unexpectedly succeeded")
    return 0


def wait_for_marker(client: CmuxClient, surface: int, marker: str) -> None:
    deadline = time.time() + 5.0
    last = ""
    while time.time() < deadline:
        last = client.read_screen(surface).text
        if marker in last:
            return
        time.sleep(0.05)
    raise AssertionError(f"marker not found; last screen: {last!r}")


def next_resized(stream, surface: int, timeout: float):
    deadline = time.time() + timeout
    old_timeout = stream._conn.sock.gettimeout()
    stream._conn.sock.settimeout(timeout)
    try:
        while time.time() < deadline:
            stream._conn.sock.settimeout(max(deadline - time.time(), 0.001))
            event = next(stream)
            if event.event == "surface-resized" and event.surface == surface:
                return event
    finally:
        stream._conn.sock.settimeout(old_timeout)
    raise CmuxTimeoutError("surface-resized not observed")


def next_title_changed(stream, surface: int, title: str, timeout: float):
    deadline = time.time() + timeout
    old_timeout = stream._conn.sock.gettimeout()
    stream._conn.sock.settimeout(timeout)
    try:
        while time.time() < deadline:
            stream._conn.sock.settimeout(max(deadline - time.time(), 0.001))
            event = next(stream)
            if (
                event.event == "title-changed"
                and event.surface == surface
                and event.title == title
            ):
                return event
    finally:
        stream._conn.sock.settimeout(old_timeout)
    raise CmuxTimeoutError("title-changed not observed")


def next_attach_output(stream, timeout: float) -> None:
    deadline = time.time() + timeout
    old_timeout = stream._conn.sock.gettimeout()
    stream._conn.sock.settimeout(timeout)
    try:
        while time.time() < deadline:
            stream._conn.sock.settimeout(max(deadline - time.time(), 0.001))
            event = next(stream)
            if event.event in ("output", "resized"):
                return
    finally:
        stream._conn.sock.settimeout(old_timeout)
    raise CmuxTimeoutError("attach output not observed")


def next_topology(stream: TopologyStream, timeout: float):
    old_timeout = stream._conn.sock.gettimeout()
    stream._conn.sock.settimeout(timeout)
    try:
        return next(stream)
    finally:
        stream._conn.sock.settimeout(old_timeout)


def find_workspace_for_surface(tree, surface: int) -> int | None:
    for workspace in tree.workspaces:
        for screen in workspace.screens:
            for pane in screen.panes:
                if any(tab.surface == surface for tab in pane.tabs):
                    return workspace.id
    return None


if __name__ == "__main__":
    raise SystemExit(main())
